pkg/scheduler/DESIGN.md

```diff
@@ -0,0 +1,689 @@
+for the queries in the `store-gateway`, `ingester-and-store-gateway`, and `unknown` queues,
+but by partitioning the querier-worker connections evenly across the four queues,
+25% of connections remain "reserved" to process queries from the `ingester` queue.
+
+The primary measure of success is the servicing of the queries to the non-degraded query component,
+
+#### Modeling the Solution
+
+Again we simplify the system to two query components and four querier connections.
+Queries to the "slow" query component take 8 ticks to process while queries to the "fast" query component take 1 tick.
+
+In 64 ticks (16 each for 4 connections), the new system:
+
+- dequeues and starts processing 36 queries: 32 fast, 4 slow
+- completes processing 36 queries: 32 fast, 4 slow
+- spends 32 ticks processing the fast queries
+- spends 32 ticks processing the slow queries
+
+Compare with the `Query Processing Time Utilization with Round-Robin` results and diagram above.
+The new system allocates more query processing time to the fast queries
+and completes 8x more fast queries than the original system in the same time period.
+
+##### Diagram: Query Processing Time Utilization with Querier-Worker Queue Prioritization
+
+```mermaid
+---
+title: Query Processing Time Utilization with Querier-Worker Queue Prioritization
+config:
+  gantt:
+    displayMode: compact
+    numberSectionStyles: 2
+  theme: default
+---
+gantt
+    dateFormat ss
+    axisFormat %S
+    tickInterval 1second
+
+    section consumer-1
+        fast        :active, c1-1, 00, 1s
+        fast        :active, c1-2, 01, 1s
+        fast        :active, c1-3, 02, 1s
+        fast        :active, c1-4, 03, 1s
+        fast        :active, c1-5, 04, 1s
+        fast        :active, c1-6, 05, 1s
+        fast        :active, c1-7, 06, 1s
+        fast        :active, c1-8, 07, 1s
+        fast        :active, c1-9, 08, 1s
+        fast        :active, c1-10, 09, 1s
+        fast        :active, c1-11, 10, 1s
+        fast        :active, c1-12, 11, 1s
+        fast        :active, c1-13, 12, 1s
+        fast        :active, c1-14, 13, 1s
+        fast        :active, c1-15, 14, 1s
+        fast        :active, c1-16, 15, 1s
+
+    section consumer-2
+        fast        :active, c2-1, 00, 1s
+        fast        :active, c2-2, 01, 1s
+        fast        :active, c2-3, 02, 1s
+        fast        :active, c2-4, 03, 1s
+        fast        :active, c2-5, 04, 1s
+        fast        :active, c2-6, 05, 1s
+        fast        :active, c2-7, 06, 1s
+        fast        :active, c2-8, 07, 1s
+        fast        :active, c2-9, 08, 1s
+        fast        :active, c2-10, 09, 1s
+        fast        :active, c2-11, 10, 1s
+        fast        :active, c2-12, 11, 1s
+        fast        :active, c2-13, 12, 1s
+        fast        :active, c2-14, 13, 1s
+        fast        :active, c2-15, 14, 1s
+        fast        :active, c2-16, 15, 1s
+
+
+    section consumer-3
```

Comment:
In the worst case, I think 3/4 consumers would see slow queries? Still an improvement, though (25% time spent on fast vs. 12.5%).

Reply:
> Again we simplify the system to two query components and four querier connections.

this example is simplified to only two "lanes", two query components, so the split is 50/50, there's just 4 consumers because having 4 offered a better illustration in the previous version of this graph above.

in reality if there are all query components active then yes it's 25% but that is explained elsewhere